src/sentry/utils/arroyo_guard.py

```
@@ -0,0 +1,112 @@
+
+    """
+
+    def guard_inner(
+        cls: type[NonFinalStrategy[TStrategyPayload, TResult]]
+    ) -> type[NonFinalStrategy[TStrategyPayload, TResult]]:
+
+        class Guard(NonFinalStrategy[TStrategyPayload, TResult]):
+            def __init__(
+                self, next_step: ProcessingStrategy[TResult], *args: Any, **kwargs: Any
+            ) -> None:
+                self.__messages_carried_over: Deque[Message[TResult] | InvalidMessage] = deque()
+                self.__instance = cls(next_step, *args, **kwargs)
+                self.__inner_submit = next_step.submit
+
+                def wrapped_inner_submit(msg: Message[TResult]) -> None:
+                    try:
+                        self.__inner_submit(msg)
+                    except MessageRejected:
+                        self.__messages_carried_over.append(msg)
+                    except InvalidMessage as exc:
+                        self.__messages_carried_over.append(exc)
+
+                setattr(next_step, "submit", wrapped_inner_submit)
+                assert isinstance(next_step, ProcessingStrategy)
+
+            def submit(self, msg: Message[TStrategyPayload]) -> None:
+                if len(self.__messages_carried_over) > max_buffer_size:
+                    raise MessageRejected
+
+                self.__instance.submit(msg)
+
+            def poll(self) -> None:
+                while self.__messages_carried_over:
+                    if isinstance(self.__messages_carried_over[0], InvalidMessage):
+                        raise self.__messages_carried_over.popleft()
+
+                    try:
+                        self.__inner_submit(self.__messages_carried_over[0])
+                        self.__messages_carried_over.popleft()
+                    except MessageRejected:
+                        break
+                    except InvalidMessage as exc:
+                        self.__messages_carried_over.append(exc)
```

if we `append` `InvalidMessage` to the right side of the queue, and then pop from the left, we might delay the re-raising of `InvalidMessage` for a while. I'm not sure but I think this means we might re-raise `InvalidMessage` when the referred message has already been committed, at which point we can't DLQ it anymore. Is there a reason to not let the exception bubble up directly?

---

an invalid message from downstream by definition must be earlier offset than a message that hasn't even been submitted yet, so technically it's always safe to reraise those first
though putting it in a buffer rather than straight reraising is supposed to solve this problem:
```
for msg in messages_for_next_step:
   inner_step.submit(msg)
```
and potentially getting an invalid message from some of these calls

---

> so technically it's always safe to reraise those first

but you aren't. your queue has offsets like this (as in `wrapped_inner_submit` you append to the right):

```
1, 2, 3, 4, 5
```

you use `popleft`, so you submit 1, 2 to next_step, and next_step raises that 2 is invalid. so now you `append` on the right:

```
3, 4, 5, invalid(2)
```

on the next poll, you forward 3, 4, 5 to next_step, and next_step commits 5.

you then raise InvalidMessage(2), but 5 has already been committed.

tldr I think you want `appendleft` here, but i'm not confident that'll be sufficient

---

sorry yeah i know this isn't happening now - i think the most correct way to achieve this is to actually maintain a separate queue for the invalid message exceptions so they are also in proper order and ensure that's empty first before processing anything in `self.__messages_carried_over`. i can add that if doing it that way makes sense to you?

---

i think that makes sense